### ISSUE TITLE
Prevent NuGetFrameworkException from nomination API

### ISSUE THREAD

**Issue by zivkan** [NuGet/Home#8385]
Since d16.1, when the project system calls nomination the framework moniker provided to us when the project's TargetFramework is invalid, is `_`, rather than `Unsupported`, as previously. This causes a `NuGetFrameworkException` when `GetShortFolderName()`. Project system is not handling exception in nomination and just reports it as a fatal error which causes a yellow bar at the top of Visual Studio and also Watson errors.

We need to report errors in the assets file, but there are certain kinds of restore errors that today prevent the assets file from being generated. This work is just to prevent the watson bugs and yellow bar, and to make VS 16.x work like 15.9 does in the same situation (an error is still listed in the error list, so customers still get feedback that something is wrong). #7717 is the long term fix.

**Comment by davkean** [comment]
I tried this in 15.9, and it repos:

``` XML
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>_,Version=v2.0</TargetFramework>
  </PropertyGroup>

</Project>
```

```
System.AggregateException: One or more errors occurred. ---> NuGet.Frameworks.FrameworkException: Invalid framework identifier ''.
   at NuGet.Frameworks.NuGetFramework.GetShortFolderName(IFrameworkNameProvider mappings)
   at NuGet.SolutionRestoreManager.VsSolutionRestoreService.<>c.<ToPackageSpec>b__34_0(TargetFrameworkInformation tfi)
   at System.Linq.Enumerable.WhereSelectArrayIterator`2.MoveNext()
   at System.Linq.Buffer`1..ctor(IEnumerable`1 source)
   at System.Linq.Enumerable.ToArray[TSource](IEnumerable`1 source)
   at NuGet.SolutionRestoreManager.VsSolutionRestoreService.ToPackageSpec(ProjectNames projectNames, IVsProjectRestoreInfo projectRestoreInfo)
   at NuGet.SolutionRestoreManager.VsSolutionRestoreService.ToDependencyGraphSpec(ProjectNames projectNames, IVsProjectRestoreInfo projectRestoreInfo)
   at NuGet.SolutionRestoreManager.VsSolutionRestoreService.NominateProjectAsync(String projectUniqueName, IVsProjectRestoreInfo projectRestoreInfo, CancellationToken token)
   at Microsoft.VisualStudio.ProjectSystem.VS.NuGet.PackageRestoreInitiator.PackageRestoreInitiatorInstance.<>c__DisplayClass15_0.<<NominateProject>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.VisualStudio.ProjectSystem.CommonProjectSystemTools.Rethrow(Exception ex)
   at Microsoft.VisualStudio.ProjectSystem.ProjectErrorReporting.<>c__DisplayClass6_0.<SubmitErrorReport>b__0()
   at Microsoft.VisualStudio.ProjectSystem.ExceptionFilter.<>c__DisplayClass2_0.<Guard>g__action|0()
   at GuardMethodClass.GuardMethod(Func`1 , Func`2 , Func`2 )
   --- End of inner exception stack trace ---
---> (Inner Exception #0) NuGet.Frameworks.FrameworkException: Invalid framework identifier ''.
   at NuGet.Frameworks.NuGetFramework.GetShortFolderName(IFrameworkNameProvider mappings)
   at NuGet.SolutionRestoreManager.VsSolutionRestoreService.<>c.<ToPackageSpec>b__34_0(TargetFrameworkInformation tfi)
   at System.Linq.Enumerable.WhereSelectArrayIterator`2.MoveNext()
   at System.Linq.Buffer`1..ctor(IEnumerable`1 source)
   at System.Linq.Enumerable.ToArray[TSource](IEnumerable`1 source)
   at NuGet.SolutionRestoreManager.VsSolutionRestoreService.ToPackageSpec(ProjectNames projectNames, IVsProjectRestoreInfo projectRestoreInfo)
   at NuGet.SolutionRestoreManager.VsSolutionRestoreService.ToDependencyGraphSpec(ProjectNames projectNames, IVsProjectRestoreInfo projectRestoreInfo)
   at NuGet.SolutionRestoreManager.VsSolutionRestoreService.NominateProjectAsync(String projectUniqueName, IVsProjectRestoreInfo projectRestoreInfo, CancellationToken token)
   at Microsoft.VisualStudio.ProjectSystem.VS.NuGet.PackageRestoreInitiator.PackageRestoreInitiatorInstance.<>c__DisplayClass15_0.<<NominateProject>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.VisualStudio.ProjectSystem.CommonProjectSystemTools.Rethrow(Exception ex)
   at Microsoft.VisualStudio.ProjectSystem.ProjectErrorReporting.<>c__DisplayClass6_0.<SubmitErrorReport>b__0()
   at Microsoft.VisualStudio.ProjectSystem.ExceptionFilter.<>c__DisplayClass2_0.<Guard>g__action|0()
   at GuardMethodClass.GuardMethod(Func`1 , Func`2 , Func`2 )<---
```

**Comment by zivkan** [comment]
@davkean I'd argue that's a different scenario as the csproj is explicitly specifying the `_` TFM. I don't believe customers are getting this exception because they're specifying a bad "long form" TFM, I believe they're  typoing a "short form" TFM.

Using this csproj:

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>zivkan2.0</TargetFramework>
  </PropertyGroup>

</Project>
```

In 15.9, NuGet's nomination receives `Unsupported,Version=0.0`, whereas in 16.3 it gets `_,Version2.0`. I don't know if it's the project system, the .NET Core SDK, or something else that changed, but something has changed the TFM that's being nominated to NuGet.

As I said in #7717, NuGet needs to be more resilient, but given that something other than NuGet has changed which lead to this exception being thrown, that component may have the opportunity to change back to the old behaviour more quickly than NuGet is changed.

**Comment by davkean** [comment]
Let me restate some of things I've been saying on this thread and other threads:

1. The project system does not manipulate the data passed to NuGet, we are not passing "Unsupported". We are passing you the raw string. I pointed it out [here](https://github.com/NuGet/Home/issues/7717#issuecomment-515638656), but I went back and confirmed it for 15.9[1]:
2. NuGet should never throw an exception back to the project-system on bad data. This is being tracked by https://github.com/NuGet/Home/issues/7717.
3. We deliberately changed from nominating TargetFramework -> TargetFrameworkMoniker as the latter represents the canonical target framework, the former does not.
4. NuGet is failing to parse the well formed target framework; `_, Version=v4.5`. While it doesn't represent any existing framework, it is well formed and should be treated as though it encounter something like `FooBar, Version=v4.5`.
5. NuGet has the same bug in 15.9 where it can fail to parse the `TargetFramework` as shown above.

[1]
![image](https://user-images.githubusercontent.com/1103906/62086344-5d853300-b2a1-11e9-93f2-1231fdf0c6e3.png)]
